### ISSUE TITLE
如何在spikingjelly库里使用bntt操作

### ISSUE THREAD

**Issue by suncy11** [fangwei123456/spikingjelly#621]
你好，想请问一下如何在spikingjelly库里执行bntt操作呢，也就是在每个时间步长都分配一个bn?

<img width="703" alt="Image" src="https://github.com/user-attachments/assets/9342ebf5-db15-435b-87d8-ea8bd0f35f51" />

**Comment by NneurotransmitterR** [comment]
可以自己写个类，继承`base.MemoryModule`，将BNTT包装为一个网络层。
```python
from spikingjelly.activation_based import base


class BatchNorm2dThroughTime(base.MemoryModule):
    def __init__(self, num_features, num_steps, step_mode='s', **kwargs):
        super().__init__()
        self.num_steps = num_steps
        self.step_mode = step_mode
        self.bn_list = nn.ModuleList([nn.BatchNorm2d(num_features, **kwargs) for _ in range(num_steps)])
        for bn in self.bn_list:
            bn.bias = None
        self.register_memory('step', 0)

    def single_step_forward(self, x):
        assert 0 <= self.step <= self.num_steps
        out = self.bn_list[self.step](x)
        self.step += 1
        return out

# 简单测试一下
x = torch.randn([2, 3, 32, 32])
x_seq = x.repeat(4, 1, 1, 1, 1)
bn2dtt = BatchNorm2dThroughTime(num_features=3, num_steps=4, step_mode='s')
out = functional.multi_step_forward(x_seq, bn2dtt)
functional.reset_net(bn2dtt)
functional.set_step_mode(bn2dtt, 'm')
y_seq = bn2dtt(x_seq)
print(torch.allclose(out, y_seq))
```

**Comment by suncy11** [comment]
for epoch in range(start_epoch, args.epochs):
        start_time = time.time()
        net.train()
        train_loss = 0
        train_acc = 0
        train_samples = 0
        for frame, label in train_data_loader:
            optimizer.zero_grad()
            frame = F.interpolate(frame, size=(2,image_size,image_size),mode='nearest')
就在for frame, label in train_data_loader:这一段，我是直接在github里下载当前的代码然后把spikingjelly的文件夹放在我自己的文件夹里用的，所以和你的版本可能不一样吗？前面提取数据集都是正常的，但当运行for frame, label in train_data_loader时就报错了

**Comment by suncy11** [comment]
train_set = DVS128Gesture(args.data_dir, train=True, data_type='frame', split_by='number', frames_number=args.T)
test_set = DVS128Gesture(args.data_dir, train=False, data_type='frame', split_by='number', frames_number=args.T)

    train_data_loader = DataLoader(
        dataset=train_set,
        batch_size=args.b,
        shuffle=True,
        num_workers=args.j,
        drop_last=True,
        pin_memory=True)

    test_data_loader = DataLoader(
        dataset=test_set,
        batch_size=args.b,
        shuffle=False,
        num_workers=args.j,
        drop_last=False,
        pin_memory=True)
前面定义数据集是这样的，这些都没有报错

**Comment by NneurotransmitterR** [comment]
较低版本的PyTorch (<=2.1) 不完全兼容numpy 2，如果先前安装了最新版的numpy则需要降级一下。
https://github.com/pytorch/pytorch/issues/107302#issue-1853509316

**Comment by suncy11** [comment]
numpy版本是2.0.1, 我更新numpy版本之后确实不会报错了，谢谢了。但是用之前版本的库，numpy用的是2.0.1，也是不会报这个错误的...

**Comment by NneurotransmitterR** [comment]
那就奇怪了，这里报错位置是PyTorch的代码，照理说跟SpikingJelly没关系

**Comment by suncy11** [comment]
还有个问题想请教一下，我之前自己写了一个把权重二值化的2d卷积（还有全连接和1d卷积，同理），然后用layer.SeqToANNContainer包装起来的。在迁移到新版本之后，现在卷积变成layer.Conv2d了，之前的二值化卷积的代码没有办法用了，不知道有什么办法可以在运行之前改变layer.Conv2d中的权重吗？
class IRConv2d(nn.Conv2d):

    def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True):
        super(IRConv2d, self).__init__(in_channels, out_channels, kernel_size, stride, padding, dilation, groups, bias)
        self.k = torch.tensor([10]).float().to(device)
        self.t = torch.tensor([0.1]).float().to(device)

    def forward(self, input):
        w = self.weight
        ba = input
        bw = w - w.view(w.size(0), -1).mean(-1).view(w.size(0), 1, 1, 1)
        bw = bw / bw.view(bw.size(0), -1).std(-1).view(bw.size(0), 1, 1, 1)
        bw = BinaryQuantize().apply(bw, self.k, self.t)
        # sw = torch.mean(abs(w))
        # bw = bw * sw
        #### ba = BinaryQuantize().apply(input, self.k, self.t)
        output = F.conv2d(ba, bw, self.bias,
                          self.stride, self.padding,
                          self.dilation, self.groups)
        return output

**Comment by NneurotransmitterR** [comment]
在新版本中也可以继续使用 `layer.SeqToANNContainer`

> 有什么办法可以在运行之前改变layer.Conv2d中的权重吗

或者你继承`layer.Conv2d`然后重写下`forward`方法
